### DISCUSSION TITLE
installing editable hnn-core - dlopen fail

### DISCUSSION THREAD

**Discussion by dylansdaniels** [jonescompneurolab/hnn-core#629] [Installation]
Running (intel) MacOS Venture 13.2.1

I'm running into a path issue with NEURON probably related to older installations. I'm getting 'dlopen fail' when running my 'make test', but it's also looking in the wrong location.

dlopen failed - 
dlopen(/Users/dylandaniels/Documents/github-repos/hnn-core/hnn_core/mod/x86_64/.libs/libnrnmech.0.so, 0x0002): Library not loaded: /Applications/NEURON-7.7/nrn/x86_64/lib/libnrnoc.0.dylib
  Referenced from: <959875B6-5E4B-374A-8C93-23C300F47E54> /Users/dylandaniels/Documents/github-repos/hnn-core/hnn_core/mod/x86_64/.libs/libnrnmech.0.so
  Reason: tried: '/Applications/NEURON-7.7/nrn/x86_64/lib/libnrnoc.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Applications/NEURON-7.7/nrn/x86_64/lib/libnrnoc.0.dylib' (no such file), '/Applications/NEURON-7.7/nrn/x86_64/lib/libnrnoc.0.dylib' (no such file), '/usr/local/lib/libnrnoc.0.dylib' (no such file), '/usr/lib/libnrnoc.0.dylib' (no such file, not in dyld cache)

I had everything working before trying to install the editable version of core, though in notebook I had to run sys.path.insert(0, '/usr/local/nrn/lib/python/') to get it to work (and also had to remove 'DISPLAY' from os.environ or i'd get a 'can't open DISPLAY' error from neruon)

**Comment by jasmainak** [comment]
Just to be clear, did you install Neuron using `pip` or did you use a different method?

**Reply by dylansdaniels** [reply]
several months back i deleted my old neuron installation (not pip) and installed via pip. now that i think about it, i'm not sure that /usr/local/nrn/lib/python is doing anything? i copied the solutions here https://doc.cocalc.com/howto/neuron.html to get things running in notebook before trying to install the editable version, but it may have just been solving the display issue

my neuron lives here: /Users/dylandaniels/anaconda3/envs/hnn/lib/python3.7/site-packages/neuron

**Comment by jasmainak** [comment]
I think those instructions might be outdated. Can you delete all copies of Neuron you have and do a fresh editable install of `hnn-core` in a fresh environment? So delete:

/usr/local/nrn/lib/python

and then make a new environment and try installing in there. Make sure to not touch any environment variables. It should just work out of the box without editing any environmental variables

**Reply by dylansdaniels** [reply]
okay will do!

**Reply by dylansdaniels** [reply]
It's working, huzzah! I should have known the answer would be to uninstall and start from scratch. 😓 

There we still 3 failed tests overall:

FAILED hnn_core/tests/test_drives.py::test_external_drive_times - re.error: global flags not at the start of the expression at position 15

FAILED hnn_core/tests/test_drives.py::test_add_drives - re.error: global flags not at the start of the expression at position 15

FAILED hnn_core/tests/test_extracellular.py::test_extracellular_api - AssertionError: Regex pattern did not match.

**Reply by jasmainak** [reply]
Hahaha :) 

The first two errors are strange. What version of Python are you using in the new environment? Can you share the full traceback?

**Reply by dylansdaniels** [reply]
Sure! I'm using v3.11.2
[maketest.txt](https://github.com/jonescompneurolab/hnn-core/files/11113952/maketest.txt)

**Reply by rythorpe** [reply]
I've never seen those errors before either. After a little digging, I suspect this is caused by a discrepancy between how Python 3.11 handles global regex flags (which have now been deprecated and now produce an error instead of a warning) and your version of pytest. Try upgrading your pytest module.

**Reply by rythorpe** [reply]
*clarification: global regex flags that are not placed at the start of an expression have been deprecated. I don't think we use these explicitly in our code, but I think the matching functionality in `pytest.raises(..., match="blah")` tries matching the contents of stderr (which might contain a global regex flag character that hasn't been escaped) to `"blah"`.

**Reply by dylansdaniels** [reply]
Huh, I'm on the latest release of pytest already (7.2.2)... let me try downgrading python instead, unless you have another suggestion?

**Reply by rythorpe** [reply]
Are you able to run the examples?

**Reply by dylansdaniels** [reply]
I've only run the 'basic GUI usage' via notebook so far (it runs just fine). optimize_evoked is running now appears to work. I can run through the other examples to see if they throw any errors if that would be helpful

**Reply by jasmainak** [reply]
This seems like a Python 3.11 issue. Instead of downgrading, I wonder if this could be your first PR @dylansdaniels to fix these tests :)

**Reply by dylansdaniels** [reply]
<img style="height:200px" src="https://www.postcardmania.com/wp-content/uploads/blog/img/its-go-time.png">

**Reply by dylansdaniels** [reply]
Also, all of the workflows run, but the plot wasn't showing correctly in the final example of the alpha workflow where you have the 3 subplots. It was drawing the first plot from plot_spikes_hist and not drawing the final fig after adding the dipole and psd subplots. 

It displays properly when adding the show=False arg to plot_spikes_hist. Just wanted to flag that

**Reply by rythorpe** [reply]
Oh, good catch. Would you mind opening an issue for that?

**Reply by dylansdaniels** [reply]
Sure!

**Comment by jasmainak** [comment]
Also I hope you don't have `NEURONHOME` or `PYTHONPATH` or any such environment variables already set in your `.bashrc` or `.bash_profile` ?

**Reply by dylansdaniels** [reply]
.bashrc is empty and .bash_profile has the following but it's commented out:

unset __conda_setup
#<<< conda initialize <<<
#export PYTHONPATH=/Applications/NEURON-7.7/nrn/lib/python:$PYTHONPATH
#export PATH=/Applications/NEURON-7.7/nrn/x86_64/bin:$PATH
